File: Sample/CSharp/Sample/Code/Int3.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using Scorpio.Commons;
using Scorpio.Table;
namespace ScorpioProtoTest {
public class Int3 : IData {
    private bool m_IsInvalid;
    private ReadOnlyCollection<Int2> _Value1;
    /// <summary> () </summary>
    public ReadOnlyCollection<Int2> getValue1() { return _Value1; }
    private int _Value2;
    /// <summary> () </summary>
    public int getValue2() { return _Value2; }
    public object GetData(string key ) {
        if (key == "Value1") return _Value1;
        if (key == "Value2") return _Value2;
        return null;
    }
    public bool IsInvalid() { return m_IsInvalid; }
    private bool IsInvalid_impl() {
        if (!TableUtil.IsInvalid(this._Value1)) return false;
        if (!TableUtil.IsInvalid(this._Value2)) return false;
        return true;
    }
    public static Int3 Read(ScorpioReader reader) {
        Int3 ret = new Int3();
        {
            int number = reader.ReadInt32();
            List<Int2> list = new List<Int2> ();
            for (int i = 0;i < number; ++i) { list.Add(Int2.Read(reader)); }
            ret._Value1 = list.AsReadOnly();
        }
        ret._Value2 = reader.ReadInt32();
        ret.m_IsInvalid = ret.IsInvalid_impl();
        return ret;
    }
}
}

File: Backup/ts/src/Code/Int3.ts
import { IScorpioReader } from '../ScorpioProto/Commons/IScorpioReader'
import { ScorpioUtil } from '../ScorpioProto/Commons/ScorpioUtil'
import { TableUtil } from '../ScorpioProto/Table/TableUtil'
import { IData } from '../ScorpioProto/Table/IData'
import { ITable } from '../ScorpioProto/Table/ITable'
import { Int2 } from './Int2'

export class Int3 implements IData {
    private m_IsInvalid:boolean = false;
    
    private _Value1:any = null;
    /*   默认值() */
    getValue1():Array<Int2> { return this._Value1; }
    ID():Array<Int2> { return this._Value1; }
    private _Value2:any = null;
    /*   默认值() */
    getValue2():number { return this._Value2; }
    
    GetData(key:string):any {
        if ("Value1" == key) return this._Value1;
        if ("Value2" == key) return this._Value2;
        return null;
    }
    
    IsInvalid():boolean { return this.m_IsInvalid; }
    private CheckInvalid():boolean {
        return false;
    }
    
    Set(value:Int3) {
        this._Value1 = value._Value1;
        this._Value2 = value._Value2;
    }
    
    ToString() {
        return `{
            Value1 : ${ScorpioUtil.ToString(this._Value1)}, 
            Value2 : ${this._Value2}
            }`;
    }
    
    public static Read(fileName:string, reader:IScorpioReader):Int3 {
        var ret = new Int3();
        {
            let list = new Array<Int2>();
            let number = reader.ReadInt32();
            for (let i = 0; i < number; ++i) { list.push(Int2.Read(fileName, reader)); }
            ret._Value1 = list;
        }
        ret._Value2 = reader.ReadInt32();
        ret.m_IsInvalid = ret.CheckInvalid();
        return ret;
    }
}